Add template feature

joshuacoles opened the pull request:
Implements a variant of the template functionality discussed in #1. Negation was added for ease of use by allowing inheritance with exceptions. Any advice welcome especially with the parsing as I have little experience in that area.

joshuacoles committed:
Implement template functionality.

Implements a variant of the template functionality discussed in #1 under the -t/--template flag. Note tests are not currently included

sindresorhus commented:
Nice! :)

I only have time for a quick look now. Will review more thoroughly in a few days.

Chalk already supports nesting, so couldn't we just replace e.g. `{bold.red foo{blue bar}}` with `chalk.bold.red('foo', chalk.blue('bar'))`?

// @Qix-

Qix- commented:
Woo, thanks for the PR @joshuacoles. Same here - only a quick look.

Could you add tests/updated README docs? I know those are the mundane parts but it'll really help validate this a bit.

joshuacoles commented:
Ahh yeah forgot about them! Would the nested `chalk.<style>(<content>)` method be preferred over the current one? The only issue I can see would be negation styles but they were just an extra and could probably be implemented in other ways if needed.

Qix- commented on cli.js:
```diff
@@ -31,12 +37,27 @@ function init(data) {
+				console.log(template(cli.flags.t));
+			} catch (x) {
+				console.error('Something went wrong! Maybe review your syntax?\n');
+				console.error(x);
```
``` javascript
console.error(x.stack);
```

Qix- commented on cli.js:
```diff
@@ -31,12 +37,27 @@ function init(data) {
+				process.exit(1);
+			}
+		} else {
+			console.log('The -t option only takes 1 argument');
```
`console.error(...)`

Qix- commented on package.json:
```diff
@@ -54,6 +54,7 @@
     "chalk": "^1.0.0",
     "dot-prop": "^2.2.0",
     "get-stdin": "^5.0.1",
+    "lodash": "^4.11.1",
```
Do we really want to pull in lodash for it being used once? Seems like a lot of overhead for a single function call. There is already `array-without` which would be much more lightweight.

joshuacoles replied:
I figured I could as it seemed like it was being pulled in any way by another thing. Should I move to the per function packages or just re-write them locally?

Qix- replied:
Size is a concern, and if another package is using a different version of lodash then they'll be fragmented anyway. Can't always count on a dependency upstream using the exact same lodash.

sindresorhus commented:
> Ahh yeah forgot about them! Would the nested `chalk.<style>(<content>)` method be preferred over the current one?

No, I meant that you might be able to translate the template tags directly to chalk calls internally.

joshuacoles commented:
Ok so I've fixed all the errors @Qix- pointed out and have transitioned to array-without over lodash. I have added tests however in the process of doing this discovered that the escaping is broken so I will fix them before I commit.

With regards to what you mentions @sindresorhus. The way I currently do it is I take all the styles applied to a chunk (e.g. `{blue.bold hey {!bold.underline you}}` would be two chunks with the styles `['blue', 'bold']` applied to 'hey' and `['blue', 'bold', '!bold', 'underline']` applied to 'you') and then apply negations (so removing the 'bold' and '!bold' from the styles applied to 'you') and then use dotProp to fetch the chalk style and then apply it. Note there are also resets surrounding each chunk as they are technically styled independently of each other when flattened and this stops bleeding.

What you are proposing, tell me if I get this wrong, is taking the style as a string as using dotProp directly on that with each sub-chunk being an argument to its parent. This would produce less redundant output as currently each style is reapplied for each chunk and resets surround each one and possibly simpler code. However I cannot see an easy way to implement negation styles with this so we would lose them (but to be fair they were just a thing I added because they were cool).

So your choice which one you'd prefer and I'll re-write as needed.

Qix- commented:
Something to think about here - `!` is a special character in a lot of unix shells. Perhaps another character is better (vote is on `~` since it literally means 'not').

I also wonder the use case here - I think the negation functionality _would_ be used if it were available but  it's not required for templating to work.

I could see it go either way. It'd be great if we could find a non-intrusive way to keep it.

Great work so far by the way @joshuacoles 💃

joshuacoles committed:
Fix errors as pointed out by @Qix-; replace lodash with array-without; add tests and readme and fix bugs regarding not requiring blocks to closed and escaping.

joshuacoles commented:
So just added tests and other things as well as fixed some bugs with the parsing. True `!` is a special character so that might be an issue but then again so are `{` and `}` as far as I know so it would still require to be in quotes.

Qix- commented on readme.md:
```diff
@@ -20,9 +20,11 @@ $ chalk --help
   Usage
     $ chalk <style> ... <string>
     $ echo <string> | chalk <style> ...
+	$ chalk -t/--template {<style> ... <string>}
```
Space indent :)

Qix- commented on readme.md:
```diff
@@ -32,6 +34,14 @@ See [supported styles](https://github.com/chalk/chalk#styles).
+- `{red.bold unicorns} are {blue FUN!!!}`
+- {red.bold unicorn {blue.underline dancing}}
+- {red red {!red normal}}
+- {red hey\} stil red} not red
```
`stil` -> `still`

Qix- commented:
That is correct, though I've still seen problems with `!` even in quotes. Perhaps @sindresorhus will have more insight.

joshuacoles committed:
Fix readme formatting.

joshuacoles commented:
Is there anything else that need to be done to this or is it just waiting for final review?

Qix- commented:
Just need to have @sindresorhus look over it. If he doesn't have time this week I'll step in and approve it ;)

sindresorhus commented on cli.js:
```diff
@@ -4,16 +4,22 @@ const getStdin = require('get-stdin');
 	  $ echo <string> | chalk <style> ...
+	  $ chalk -t/--template {<style> ... <string>}
 
 	Example
```
Examples

sindresorhus commented on cli.js:
```diff
@@ -31,12 +37,27 @@ function init(data) {
-		console.error('Input required');
-		process.exit(1);
-	}
+	if (cli.flags.t) {
```
``` js
if (cli.flags.template) {
```

sindresorhus commented on cli.js:
```diff
@@ -31,12 +37,27 @@ function init(data) {
+	if (cli.flags.t) {
+		if (cli.input.length === 0) {
+			try {
+				console.log(template(cli.flags.t));
```
Same as https://github.com/chalk/chalk-cli/pull/4/files#r63409284

sindresorhus commented on cli.js:
```diff
@@ -31,12 +37,27 @@ function init(data) {
+		if (cli.input.length === 0) {
+			try {
+				console.log(template(cli.flags.t));
+			} catch (x) {
```
x => err

sindresorhus commented on cli.js:
```diff
@@ -31,12 +37,27 @@ function init(data) {
+				process.exit(1);
+			}
+		} else {
+			console.error('The -t option only takes 1 argument');
```
`-t` => `--template`

takes => accepts

sindresorhus commented on package.json:
```diff
@@ -51,6 +51,7 @@
     "text"
   ],
   "dependencies": {
+    "array-without": "0.0.2",
```
Use https://github.com/sindresorhus/arr-exclude

sindresorhus commented on readme.md:
```diff
@@ -32,6 +34,14 @@ See [supported styles](https://github.com/chalk/chalk#styles).
 
 - [chalk](https://github.com/chalk/chalk) - API for this module
 
+## Template Syntax
```
add an empty line below here.

sindresorhus commented on templates.js:
```diff
@@ -0,0 +1,157 @@
+
+	let current = root;
+
+	for (let i = 0; i < string.length; i++) {
```
Use a `for...of` loop

joshuacoles replied:
The index of the character is required for backtracking when checking for escaping. I do not think this is provided by a `for ... of`. Correct me if I'm wrong.

sindresorhus replied:
You could use `for (const [i, el] of string.entries()) {`, but it seems Node.js 4 doesn't support `.entries()`, so never mind.

joshuacoles replied:
Ok will push changes without for of loop. Any verdict on negations?

sindresorhus commented:
> What you are proposing, tell me if I get this wrong, is taking the style as a string as using dotProp directly on that with each sub-chunk being an argument to its parent. This would produce less redundant output as currently each style is reapplied for each chunk and resets surround each one and possibly simpler code. However I cannot see an easy way to implement negation styles with this so we would lose them (but to be fair they were just a thing I added because they were cool).

Correct.

I honestly don't see the practical use-case for negation. I've used `chalk` programmatically for years and made a lot of CLI apps and never found the need for negation. I'm curious what problems you're trying to solve with it that can't be done without negation?

sindresorhus commented:
Sorry for the slow response. My pull request (and browser tab) backlog is a bit overwhelming.

joshuacoles commented:
Specifically it was for brevity. Suppose you had some text in bold however wanted one section to be normal. Instead of having to write `{blue.bold I'm bold} {blue I'm not} {blue.bold I am again}` you could write `{blue.bold I'm bold {!bold I'm not} I am again}` which is shorter and I think easier to read. Your call though.

joshuacoles commented:
Ok I have fixed everything but the `for...of` vs `for(let i; ...)` loop locally. I believe it requires an index based loop but if you can think of a way to implement it without I'm all ears. And don't worry about timings I'm happy to wait! :)

joshuacoles committed:
Address comments made by @sindresorhus.

joshuacoles commented:
No as the bold styling would affect the `(with added fairies)` which negation styles prevent without having to reapply the wanted styles to a new chunk.

Qix- commented:
@joshuacoles could you change the negation character to `~`?

joshuacoles committed:
Change negation character to `~`

joshuacoles commented:
@Qix- Done!

sindresorhus commented:
Ok, I misunderstood how the negation worked.

This would result in the same though, without negation:

```
chalk -t '{red.bold Dungeons and Dragons {reset.blue (with added fairies)}}'
```

Or even better and clearer:

```
chalk -t '{red.bold Dungeons and Dragons} {blue (with added fairies)}'
```

## 

I'm not totally against negation, but I have yet to see an actual convincing and realistic use-case.

sindresorhus commented on cli.js:
```diff
@@ -4,16 +4,23 @@ const getStdin = require('get-stdin');
 	Example
 	  $ chalk red bold 'Unicorns & Rainbows'
+	  $ chalk -t '{red.bold Unicorns & Rainbows}'
+	  $ chalk -t '{red.bold Dungeons and Dragons {!bold.blue (with added fairies)}'
```
This is missing a `}`

joshuacoles replied:
Fixed locally.

sindresorhus commented on readme.md:
````diff
@@ -20,9 +20,11 @@ $ chalk --help
   Example
     $ chalk red bold 'Unicorns & Rainbows'
+    $ chalk -t '{red.bold Unicorns & Rainbows}'
 ```
````
You need to sync this with the actual CLI output.

joshuacoles replied:
I don't quite understand what you mean? Do you mean the image?

joshuacoles commented:
While in that case it a reset style would work imagine a situation with `{bold.red.italic Many styles {~italic not italic} and back}` if you were to use resets it would have to be `{bold.red.italic Many styles {reset.bold.red not italic} and back}` and limits the power of nesting significantly.

Qix- commented:
I think Sindre's point is the alternative.

```
{bold.red {italic Many styles} not italic {italic and back}}
```

joshuacoles commented:
Ah ok, that makes sense. Would that be preferred then over negation then?

Qix- commented:
I'll finally take a look at the PR tonight and check it out, but if the negation logic isn't intrusive then I don't see why we shouldn't support it if someone finds an edge case where they want to use it (or think in terms of subtractive styles instead of additive). If you've already done the work and it's not going to affect regular usage I don't see a reason to _take it out_.

joshuacoles committed:
Fix formatting

sindresorhus commented:
Landed at last \o/

Thank you @joshuacoles for your hard work on this :)

## 

Sorry about the long delays. I'm usually faster at reviewing stuff.

Qix- commented:
![Yeah!](https://media.giphy.com/media/dkGhBWE3SyzXW/giphy.gif)
